workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/androidx/StateRegistryAggregator.kt

```diff
@@ -0,0 +1,300 @@
+          "This is most easily remedied by giving your container Screen rendering a unique " +
+          "Compatible.compatibilityKey, perhaps by wrapping it with Named.",
+        e
+      )
+    }
+
+    // Even if the parent lifecycle is in a state further than CREATED, new observers are sent all
+    // the lifecycle events required to catch the observer up to the current state, so this handles
+    // both the cases where we're ready to immediately restore, and where we have to wait.
+    parentLifecycle.addObserver(lifecycleObserver)
+  }
+
+  /**
+   * Must be called when the owning view detaches from the window.
+   *
+   * Stops listening to the parent lifecycle and unregisters from the parent registry.
+   */
+  public fun detachFromParentRegistry() {
+    // parentKey will only be null if parentRegistryOwner is also null.
+    parentRegistryOwner?.savedStateRegistry?.unregisterSavedStateProvider(parentKey!!)
+    parentRegistryOwner?.lifecycle?.removeObserver(lifecycleObserver)
+    parentRegistryOwner = null
+    parentKey = null
+    children.clear()
+  }
+
+  /**
+   * Puts a new [ViewTreeSavedStateRegistryOwner] in place on [view], registered
+   * with its [WorkflowLifecycleOwner]. **This method must be called before [view]
+   * is attached to a window.**
+   *
+   * Clean up requirements after making this call are nuanced:
+   *
+   *  - all instances are dropped after [detachFromParentRegistry] is called
+   *
+   *  - if [view] is dropped but may be restored before [detachFromParentRegistry] is called,
+   *    as when pushing and popping a back stack, call [saveChildRegistryOwner] to capture
+   *    its state tear down its [ViewTreeSavedStateRegistryOwner]. The saved state
+   *    will be restored to the next [view] passed to [installChildSavedStateRegistryOwnerOn]
+   *    with the same [key]
+   *
+   *  - if [view] is dropped and will not be restored, as when a window is closed or
+   *    back stack history is modified, call [prune] _with the keys of the views that
+   *    remain active_.
+   *
+   * @param key identifier for the new [ViewTreeSavedStateRegistryOwner], unique across this
+   * [StateRegistryAggregator]. Typically this is derived from the
+   * [compatibility key][com.squareup.workflow1.ui.Compatible.keyFor] of the [view]'s
+   * rendering.
+   */
+  public fun installChildSavedStateRegistryControllerOn(
+    view: View,
+    key: String
+  ) {
+    val lifecycleOwner = checkNotNull(WorkflowLifecycleOwner.get(view)) {
+      "Expected container view to have set a WorkflowLifecycleOwner on its immediate " +
+        "child $view."
+    }
+    val registryOwner = KeyedStateRegistryOwner(key, lifecycleOwner)
+    children.put(key, registryOwner)?.let {
+      throw IllegalArgumentException("$key is already in use, it cannot be used register $view")
+    }
+    ViewTreeSavedStateRegistryOwner.get(view)?.let {
+      "$view already has ViewTreeSavedStateRegistryOwner: $it"
```

Comment:
Is this supposed to be an exception? Isn't it just a naked string here?

Reply:
Yup, need to add a test covering it.